docs/architecture/data-sources.md

```diff
@@ -0,0 +1,113 @@
+dotnet add CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource/src/CarbonAware.DataSources.MyNewDataSource.csproj reference CarbonAware/src/CarbonAware.csproj
+cd CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource/src
+dotnet add package Microsoft.Extensions.DependencyInjection
+```
+
+### Adding/Extending a Data Source Interface
+Each new data source should extend from a generic data source interface. A data source interface defines all the parameters and functions that any data source that falls under it's purview must define/implement. By defining the interface, it allows the SDK to switch between the set of data sources seamlessly because they all share the same input functions and output types. 
+
+Currently there is one data source interface defined -  `ICarbonIntensityDataSource` - which is the interface for all data sources that provide carbon intensity data. A new data source interface should be defined only when there is a new general area of calculation that is being introduced to the SDK.
+
+```csharp
+using CarbonAware.Interfaces;
+using CarbonAware.Model;
+using Microsoft.Extensions.Logging;
+
+namespace CarbonAware.DataSources.MyNewDataSource;
+
+public class MyNewDataSource: ICarbonIntensityDataSource 
+{
+    ...
+}
+```
+
+### Add Dependency Injection Configuration
+
+The SDK uses dependency injection to load registered data sources based on set environment variables. For a data source to be registered, it need to have a Service Collection Extension defined. To do so, add a `Configuration` directory in your data source project and create a new ServiceCollectionExtensions file. We have provided a command snippet below:
+
+```sh
+cd src/CarbonAware.DataSources/CarbonAware.DataSources.MyNewDataSource/src
+mkdir Configuration
+touch Configuration\ServiceCollectionExtensions.cs
+```
+
+Using the skeleton below, add the data source specific configuration and implementation instances to the service collection.
+
+```csharp
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.DependencyInjection.Extensions;
+
+namespace CarbonAware.DataSources.MyNewDataSource.Configuration;
+
+public static class ServiceCollectionExtensions 
+{
+    public static void AddMyNewDataSource(this IServiceCollection services)
+    {
+        // ... register your data source with the IServiceCollection instance
+    }
+}
+```
+
+### Register the New Data Source
+Once the data source's ServiceCollectionExtensions is configured, it can be registered as an available data source for the SDK. This is done in the `CarbonAware.DataSources.Registration` project. In particular, there is an equivalent `ServiceCollectionExtensions` file in the `Configuration` directory of that project. The public function `AddDataSourceService` contains a switch statement that reads in the environment variables for the SDK and selects the requested data source. To register your new data source, you must: 
```

Comment:
Consider linking to the appropriate file somewhere in this paragraph

Reply:
What do you mean the appropriate file? Where the `AddDataSourceService` lives?

Reply:
Yes.  Rather than describing how to find it, with a link this paragraph could become:

```text
Once the data source's ServiceCollectionExtensions is configured, it can be registered as an available data source for the SDK by adding to the switch statement found in the `AddDataSourceService` function of [this file](/src/link/to/file).
```